### ISSUE TITLE
Compilation problems

### ISSUE THREAD

**Issue by o0Djeen0o** [TypeStrong/atom-typescript#93]
Hi,
I've switched up from 0.57.0 to the 0.59.0 release.

- Compilation became really slow, previously instantaneous on ctrl+s, now can easily take 20/30s first time and never end afterwards.
- Even with a compilation of the project (ctrl+shift+b) the problem still occur.

Node seems to take much CPU resources in the meanwhile.

Could you help to investigate what is going wrong with the new version?

Thanks in advance for your help

**Comment by o0Djeen0o** [comment]
After some restarts compile on save feature seems to be working again as intended, sorry for the inconvenience.

**Comment by basarat** [comment]
Two possible causes:

Invalid tsconfig file. Causes the worker to crash silently. Needs error reporting to user. 

Chokidar file watching can perform bad sometimes (historically). No idea why though.

**Comment by o0Djeen0o** [comment]
Hi,

Sadly the problem persists :
* Even while just writing (no saving) the editor freeze
* Constant high consumption of V8 in background
* No compilation at saving time
* Using the project compilation (ctrl+shift+b) doesn't really trigger new build.

I'm using an auto-created tsconfig.json file (created by the plugin) and tried each plugin up to 0.60 release. (now on atom 0.181).

It's really difficult to work on my project. Could I give you more information to help on the subject ?

```
{
    "compilerOptions": {
        "target": "es5",
        "module": "amd",
        "declaration": false,
        "noImplicitAny": false,
        "removeComments": true,
        "noLib": false
    },
    "filesGlob": [
        "./**/*.ts",
        "!node_modules/**/*.ts"
    ],
    "files": [
        "./definitions/bootstrap/bootstrap.d.ts",
        "./definitions/crossroads/crossroads.d.ts",
        "./definitions/hasher/hasher.d.ts",
        "./definitions/jasmine/jasmine.d.ts",
        "./definitions/jquery/jquery.d.ts",
        "./definitions/js-signals/js-signals.d.ts",
        "./definitions/knockout-projections/knockout-projections.d.ts",
        "./definitions/knockout/knockout-3.2.0.d.ts",
        "./definitions/knockout/knockout.d.ts",
        "./definitions/mousetrap/mousetrap-global-bind.d.ts",
        "./definitions/mousetrap/mousetrap.d.ts",
        "./definitions/sortable/sortable.d.ts",
        "./node_modules/gulp-tsc/node_modules/typescript/bin/lib.core.d.ts",
        "./node_modules/gulp-tsc/node_modules/typescript/bin/lib.core.es6.d.ts",
        "./node_modules/gulp-tsc/node_modules/typescript/bin/lib.d.ts",
        "./node_modules/gulp-tsc/node_modules/typescript/bin/lib.dom.d.ts",
        "./node_modules/gulp-tsc/node_modules/typescript/bin/lib.es6.d.ts",
        "./node_modules/gulp-tsc/node_modules/typescript/bin/lib.scriptHost.d.ts",
        "./node_modules/gulp-tsc/node_modules/typescript/bin/lib.webworker.d.ts",
        "./node_modules/gulp-tsc/node_modules/typescript/bin/typescript.d.ts",
        "./node_modules/gulp-tsc/node_modules/typescript/bin/typescriptServices.d.ts",
        "./node_modules/gulp-tsc/node_modules/typescript/bin/typescriptServices_internal.d.ts",
        "./node_modules/gulp-tsc/node_modules/typescript/bin/typescript_internal.d.ts",
        "./src/app/item.ts",
        "./src/app/router.ts",
        "./src/app/startup.ts",
        "./src/app/ui/ui.ts",
        "./src/components/base/comment/comment.ts",
        "./src/components/base/input-selector/input-selector.ts",
        "./src/components/base/input/input.ts",
        "./src/components/base/multivalue/multivalue.ts",
        "./src/components/base/select/select.ts",
        "./src/components/base/typeselection/typeselection.ts",
        "./src/components/base/value-selector/valueSelector.ts",
        "./src/components/comparators/interval-switch/parser.ts",
        "./src/components/comparators/interval-switch/switch.ts",
        "./src/components/comparators/numbers/numbers.ts",
        "./src/components/comparators/switch/switch.ts",
        "./src/components/empty-node/empty-node.ts",
        "./src/components/functions/flat/flat.ts",
        "./src/components/functions/free/free.ts",
        "./src/components/home-page/home.ts",
        "./src/components/nav-bar/nav-bar.ts",
        "./src/components/search/search.ts",
        "./src/logic/parameter.ts",
        "./src/logic/parameters.ts",
        "./src/references.d.ts",
        "./test/SpecRunner.browser.ts",
        "./test/SpecRunner.karma.ts",
        "./test/components/home-page.ts"
    ]
}
```

**Comment by basarat** [comment]
@o0Djeen0o thanks for raising this. To help us narrow out the possible causes can you completely disable chokidar. I.e. comment out the entire body of `watchProjectFile` so that we have: 

```ts
function watchProjectFile(projectFile: tsconfig.TypeScriptProjectFileDetails) {
    /*if (watchingProjectFile[projectFile.projectFilePath]) return; // Only watch once

    var watcher = chokidar.watch(projectFile.projectFilePath, { ignoreInitial: true, persistent: true });
    watchingProjectFile[projectFile.projectFilePath] = watcher;

    function clear(datPathYo) {
        // TODO : Invalidate only matching caches for projectFilePath
        // Right now: Just invalidate *all*
        projectByProjectPath = {};
        projectByFilePath = {};
    }

    watcher.on('change', clear)
        .on('unlink', clear);*/
}
```

and see if it fixes it.

**Comment by basarat** [comment]
see if `v0.63.0` fixes it. I suspect it will as it has a much more stable node worker :heart:

**Comment by o0Djeen0o** [comment]
Thank you for your response.

I've tried the 0.63.0 release but it appears that problem still occurs.

Then I've tried your suggestion and completely disabled chokidar integration. For now my problem seems not reproducible, so it appears that your assumption is correct. I will update my answer if things went wrong again.

**Comment by basarat** [comment]
note to self: if it really does boil down to `chokidar` we could just *reload* hard refresh `tsconfig` cache from disk on `save`.

**Comment by csnover** [comment]
I see some stack traces pointing to `atom-typescript` via chokidar when saving files that are not TypeScript files and not in projects that should be touched by atom-typescript. From the stack trace it looks like it’s hooked itself into several of the `fs` methods, so any time atom-typescript is active and a filesystem error occurs, it’s going to get reported as being the fault of atom-typescript. (In my case, https://github.com/atom/atom/issues/4467 is the real bug.) Seems like this is probably not…desirable.

**Comment by basarat** [comment]
Alright I've moved this to the top of my queue. Will remove the chokidar dependency and make sure not to break this  https://github.com/TypeStrong/atom-typescript/issues/83

**Comment by basarat** [comment]
pushed in `v0.67.0`. @o0Djeen0o please comment / reopen if you still have issues

**Comment by ciuncan** [comment]
hi @basarat, I think this issue persists with 5.4.2. 

This plugin was working brilliantly until I tried adding gulp build workflow. It started showing thousands of error and still counting up, also blocks and fills some process queue (either atom's or atom-typescript's, not sure). Errors are about duplicate definitions shown at the `lib.d.ts` etc. files in `node_modules/gulp-tsc/`, which I guess contains an `typescript` module dependecny in it.

Here is my tsconfig.json:
```json
{
    "compilerOptions": {
        "removeComments": false,
        "sourceMap": true,
        "target": "es5",
        "module": "commonjs",
        "outDir": "app/dist/js"
    },
    "fileGlob": "app/src/**/*.ts"
}
```
and package.json:
```json
{
  "version": "0.0.0",
  "description": "",
  "main": "Gulpfile.js",
  "dependencies": {},
  "devDependencies": {
    "gulp": "^3.9.0",
    "gulp-plumber": "^1.0.1",
    "gulp-sass": "^2.0.4",
    "gulp-tsc": "^1.1.0",
    "gulp-uglify": "^1.2.0"
  }
}
```

Am I doing something wrong? Is there a recommended way to have typescript compilation in gulp build process?

**Comment by basarat** [comment]
@ciuncan if yo can provide the repo I'd be happy to debug it deeply. Also feel free to create a new issue specific to you problem 🌹

**Comment by ciuncan** [comment]
@basarat Thank you for your concern. This is a bit embarrassing but tried re-adding "gulp-tsc" back (prior to that, I have switched backed to atom-typescript compilation via tsconfig.json) but I couldn't reproduce the problem. So good news is it now works fine. I did a clean atom reinstall including `$HOME/.atom` maybe it did the trick. 

If such a thing occurs again, I'd be coming back with concrete examples so it could be debugged.

Thanks again.
